src/shared/components/copyDownloadControls/CopyDownloadControls.tsx

```diff
@@ -31,77 +44,154 @@ export class CopyDownloadControls extends React.Component<ICopyDownloadControlsP
-                        mouseLeaveDelay={0}
-                        mouseEnterDelay={0.5}
-                        placement="top"
-                        arrowContent={arrowContent}
-                    >
-                        <Button className="btn-sm" onClick={this.handleDownload}>
-                            <i className='fa fa-cloud-download'/>
-                        </Button>
-                    </DefaultTooltip>
-                </If>
-            </ButtonGroup>
+                            <button
+                                ref={(el:HTMLButtonElement) => {this._copyButton = el;}}
+                                className="btn btn-sm btn-default"
+                                data-clipboard-text="NA"
+                                id="copyButton"
+                            >
+                                <i className='fa fa-clipboard'/>
+                            </button>
+                        </DefaultTooltip>
+                    </If>
+
+                    <If condition={this.props.showDownload}>
+                        <DefaultTooltip
+                            overlay={<span>Download CSV</span>}
+                            mouseLeaveDelay={0}
+                            mouseEnterDelay={0.5}
+                            placement="top"
+                            arrowContent={arrowContent}
+                        >
+                            <Button className="btn-sm" onClick={this.handleDownload}>
+                                <i className='fa fa-cloud-download'/>
+                            </Button>
+                        </DefaultTooltip>
+                    </If>
+                </ButtonGroup>
+                <Modal
+                    show={this.downloadingData}
+                    onHide={() => undefined}
+                    bsSize="sm"
+                    className={`${copyDownloadStyles["centered-modal-dialog"]}`}
+                >
+                    <Modal.Body>
+                        <ThreeBounce style={{ display:'inline-block', marginRight:10 }} />
+                        <span>Downloading Table Data...</span>
+                    </Modal.Body>
+                </Modal>
+                <Modal
+                    show={this.copyingData}
+                    onHide={() => undefined}
+                    onEntered={() => {
+                        this.bindCopyButton(this._modalCopyButton, this._modalCopyButtonContainer);
+                    }}
+                    bsSize="sm"
+                    className={`${copyDownloadStyles["centered-modal-dialog"]}`}
+                >
+                    <Modal.Header>
+                        Download Complete!
+                    </Modal.Header>
+                    <Modal.Body>
+                        Please click on Copy to copy downloaded data to clipboard.
+                    </Modal.Body>
+                    <Modal.Footer>
+                        <span
```

Comment:
The button needs to be wrapped with another HTML element, in order for the clipboard to work in a Modal: https://stackoverflow.com/questions/38398070/bootstrap-modal-does-not-work-with-clipboard-js-on-firefox